### ISSUE TITLE
Getting ready for 3.0.0

### ISSUE THREAD

**Issue by RolfSander** [KineticPreProcessor/KPP#50]
Let's list all the remaining little tasks here which we'd like to finish before version 3.0.0 is released:

- [x] Status of Matlab?

- [x] Status of C?

- [x] duplicate Table numbers in pdf of user manual

- [x] Add to docs: How to use `F90_RCONST` for including external functions like `Hg_RateLawFuncs`

- [x] Add to docs: Mention in KPP revision history that arr, arr2, and
      k_3rd* were deleted and replaced by arr_* and k3rd*.

- [x] Maybe add to docs: Short autoreduce description in Numerical methods -> Rosenbrock methods?

- [x] I think that `DOUBLE COMPLEX` does not exist in f90 and must be replaced by `COMPLEX(kind=dp)`.

- [x] Update the list of ci-tests (_table-ci-tests) in 06_info_for_kpp_developers.rst (@yantosca?)

- [x] Add a few lines about feuler.f90 in 07_numerical_methods.rst (@msl3v?)

- [x] Cleanup of CONTRIBUTING.rst and SUPPORT.rst: Reduce overlap, maybe one file is sufficient. Sync with CONTRIBUTING.md and SUPPORT.md. (I can take care of this if you want)

- [x] Delete NFLUX completely?

- [x] Is Vdotout unnecessary?

- [x] bibtex.json seems to be outdated (or unnecessary?)

- [x] Do we have to fix these memory leaks in gen.c? https://github.com/KineticPreProcessor/KPP/compare/main...kilicomu:KPP:bugfix/gen_memory_leaks

- [x] Update doc for MacOS-specific configuration edits

- [x] Removed host-specific tests in `Makefile.defs`; added `KPP_FLEX_LIB_DIR` to specify the path of the flex library file (`libfl.a` or `libfl.so`) if it cannot be found in standard locations such as `/usr/lib`, `/usr/lib64/`, etc.

**Comment by yantosca** [comment]
Was playing around with the C-language output (tried to build radau90 in C).  Got this error:

```console
KPP is generating the ODE function:
*** Error in `kpp': free(): invalid next size (normal): 0x000000003355d390 ***
======= Backtrace: =========
/lib64/libc.so.6(+0x81329)[0x2ab6bfcb4329]
```
Looks like it is a pointer issue -- trying to free a pointer that isn't associated.

**Comment by yantosca** [comment]
Making a few minor updates to the C-language output.  Will push a PR later.

**Comment by jimmielin** [comment]
Thanks @RolfSander for starting this. I'll add to docs some description about the auto-reduction - opening a PR shortly.

I would also suggest to expand
> How to use F90_RCONST for including external functions like Hg_RateLawFuncs

with some guidance on leveraging `F90_GLOBAL` to inline code that is used to interface KPP with models. `State_Het` to manage state information for heterogeneous chemistry in GEOS-Chem comes to mind. While this is not a KPP 3.0.0 feature for all models, it's a pretty important development in GEOS-Chem with KPP to save memory and repeat computation. I'm happy to write a blurb about this and @yantosca could review it for adding to the documentation if that's okay.

**Comment by yantosca** [comment]
That's fine, @jimmielin.  Happy to review.

**Comment by RolfSander** [comment]
I like the idea to call our version a "release candidate" when we submit the manuscript!

**Comment by yantosca** [comment]
@jimmielin let me see if I can fix that setting

**Comment by yantosca** [comment]
@jimmielin @RolfSander I fixed the ReadTheDocs setting.  Latest points to dev and stable always points to the latest release.

**Comment by yantosca** [comment]
@jimmielin: I often preview the ReadTheDocs on my laptop with sphinx-autobuild.  I can tell you more about that offline if you want.

**Comment by yantosca** [comment]
> I like the idea to call our version a "release candidate" when we submit the manuscript!

We can tag it with 3.0.0-rc.0

**Comment by jimmielin** [comment]
Thanks @yantosca! This will be very helpful so `dev` documentation can be quickly viewed. Once all the items here are finished we can tag `3.0.0-rc.0` for manuscript review, then release a final `3.0.0` once everything is cleared with the reviewers.

**Comment by RolfSander** [comment]
Yes: `3.0.0-rc.0`

**Comment by RolfSander** [comment]
@yantosca: Thanks, pushing updated documentation to the dev branch also
updates the readthedocs page now!

Creating a local preview with sphinx-autobuild also works fine for me.
However, the result is slightly different (I have an older sphinx
version), so it's good to have both options.

**Comment by yantosca** [comment]
Also I went on readthedocs.org (the configuration page) and was able to activate the 2.6.0 and 2.5.0 RTD builds.  When you click on the v.latest tab at bottom left, the older versions of the docs are visible as well.

**Comment by RolfSander** [comment]
Nice to have the docs of other versions as well! I encountered a couple
of small problems with RTD. Nothing terribly important but if you know
how to fix it, that would be great:

- The references are not sorted alphabetically. I tried changing
  `default_sort_style` in conf.py but it didn't work.

- Table numbers are duplicate in the pdf, e.g., "Table 22: Table 18.
  Continuous integration tests". I think we can delete all manually
  defined table numbers in the *.rst files.

**Comment by yantosca** [comment]
@RolfSander I can look into the references.  I think the default is to sort by the order of citation but there must be ways to change that.

I'll also go thru and remove the table numbers.

**Comment by yantosca** [comment]
I also removed the table numbers from table titles.

**Comment by RolfSander** [comment]
Thanks, @yantosca! I had also tried alpha instead of unsrt but it didn't work. I must have done something wrong, I guess...

Thanks also for adjusting the BibTeX keys! The new keys are much better for the alphabetically sorted list of references.

**Comment by RolfSander** [comment]
Thanks for removing the table numbers. I deleted 3 more, I think they're all gone now.

**Comment by RolfSander** [comment]
I've finished copy-editing of the whole documentation now. A few remaining
tasks are added to the to-do list at the top of this page.

**Comment by jimmielin** [comment]
> Cleanup of CONTRIBUTING.rst and SUPPORT.rst: Reduce overlap, maybe one file is sufficient. Sync with CONTRIBUTING.md and SUPPORT.md. (I can take care of this if you want)

Maybe we could just point to the documentation and remove the `.md` files. Making a symbolic link from `/CONTRIBUTING.md` to `/docs/CONTRIBUTING.rst` could also work, since `.rst` files also show up nicely in GitHub.

**Comment by msl3v** [comment]
I can add feuler.f90 words. Anything else?

**Comment by RolfSander** [comment]
@msl3v: Maybe add a one-liner for ComputeFamilies in table-util-fun in 05_output_from_kpp.rst?

**Comment by msl3v** [comment]
I don't think ComputeFamilies() is used, nor is it functional. I think 
it's an artifact from early families development.
All of the reactions associated with a given family definition get the 
same variable, so the 'sum' that the ComputeFamilies() routine was 
originally performing is no longer necessary.

( Looking back in the repo, I had the call to GenerateComputeFamilies() 
commented out. Not sure when it was brought back to life. see back to 
commit b9b95f1a30e6a8b7d0a736813ae60ded61cb2fcc )

**Comment by yantosca** [comment]
I updated the C-I test list and also now link to the `CONTRIBUTING.md` and `SUPPORT.md` in the root folder.

**Comment by RolfSander** [comment]
Thanks, @yantosca!

Can you help me to understand how this works now? In index.rst, I still see SUPPORT.rst and CONTRIBUTING.rst. How does sphinx know that it has to use CONTRIBUTING.md and SUPPORT.md instead?

**Comment by RolfSander** [comment]
@msl3v, @yantosca: Is this line still needed in `gen.c`?

```FLUX = DefvElm( "FLUX", real, -NFLUX, "Captured flux through reactions (global)" );```

AFAICS, `NFLUX` is undefined.

**Comment by msl3v** [comment]
@RolfSander: None of the FLUX ops are needed. I will clean that all out 
now and push.

**Comment by yantosca** [comment]
@RolfSander @jimmielin: I think the purpose of `bibtex.json` is to make sure that only the references cited there show up in the references output.  That might be handy if you are using a `.bib` file with many references.  Maybe in our case we don't need it though.

**Comment by yantosca** [comment]
@RolfSander, @jimmielin: I would fix the memory leaks.  That might be important if users have several big mechanisms.  I can work on adding that in.  

The newer versions of GCC have options to check for memory leaks so that you don't have to use an external tool like Valgrind.  See: https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html

**Comment by RolfSander** [comment]
@yantosca: The values inside Vdotout are certainly important, I don't
doubt that. I just noticed that SUBROUTINE Fun already returns exactly
these data via `Vdot` even if you don't supply the optional argument
Vdotout!

**Comment by jimmielin** [comment]
Thanks @yantosca, fixing the memory leaks is definitely important. Thanks for fixing the CI tests as well together with @laestrada :)

I'm wrapping up the manuscript right now. It looks like we can get `3.0.0.rc.1` out this week or next. I'll send out the updated version for a final round of comments and we should be ready to go if there aren't any blocking bugs before release :-)

**Comment by RolfSander** [comment]
@yantosca: Thanks for fixing the C-I tests!

**Comment by jimmielin** [comment]
Looks like we are using `sphinx==3.5.4` so we should not need `bibtex.json` anymore: https://github.com/KineticPreProcessor/KPP/blob/dev/docs/requirements.txt

**Comment by RolfSander** [comment]
@jimmielin: Yes, creating the release candidate version soon is fine
with me. We can continue to work on the tasks in this issue while the
manuscript is under review.

**Comment by RolfSander** [comment]
Sorry, I meant the sphinxcontrib-bibtex version, not the sphinx version. Still, with 2.2.0 it's fine for us to delete `bibtex.json`.

**Comment by yantosca** [comment]
We can delete bibtex.json.  I'll do it.

I also manually added @kilicomu's updates into dev (they were only a few lines)

**Comment by yantosca** [comment]
I just now pushed some doc for MacOS X.

**Comment by jimmielin** [comment]
Hi @yantosca, it looks like `Vdotout` can probably be removed, if there are no more bindings to it in GEOS other than `fullchem_mod.F90`. I looked at the `dev` branch and the only place we use `Vdotout` is:

```fortran
          CALL Fun( C(1:NVAR), C(NVAR+1:NSPEC), RCONST,                          &
                    Vloc,      Aout=Aout,       Vdotout=Vdotout )
          NOxTau = Vdotout(ind_NO) + Vdotout(ind_NO2) + Vdotout(ind_NO3)         &
                 + 2.*Vdotout(ind_N2O5) + Vdotout(ind_ClNO2) + Vdotout(ind_HNO2) &
                 + Vdotout(ind_HNO4)
```

Some other calls get `Vdotout` but they're not used anywhere (and it's a local variable so it doesn't propagate to other places). It can probably be replaced with `Vloc` which has the same value as `Vdotout`, and we can get rid of an `OPTIONAL` in `Fun()`:

We don't have to remove this right away in KPP 3.0.0.rc.1 - I'll open a PR for 3.0.0 to get rid of `Vdotout` and we can do the changes in GEOS-Chem downstream first then merge into KPP later.

**Comment by kilicomu** [comment]
@yantosca Just to say that I haven't exactly extensively tested those fixes! Hence not PRing them yet.

**Comment by RolfSander** [comment]
@kilicomu: Thanks for letting us know!

Have you also seen the MCM-related discussion we're having here: https://github.com/KineticPreProcessor/KPP/issues/4 ? It would be great if we can collaborate in improving the MCM-generated KPP files so that they can be used out-of-the-box :-)

**Comment by RolfSander** [comment]
A new question: Should we rename the "map file" to "log file"?

The name "map file" is normally used for color maps and for a couple of
other purposes, see: https://en.wikipedia.org/wiki/MAP_(file_format)

The KPP-generated "map file", however, is in reality a logfile with
information about the current execution of KPP. I suggest to change the
file suffix from `.map` to `.log`.

**Comment by yantosca** [comment]
@RolfSander I like this idea.  We'd also need to update all the Makefile distclean targets as well.

**Comment by RolfSander** [comment]
Good point! I'll adjust the Makefiles as well.

**Comment by RolfSander** [comment]
I don't think we should wait with submission of the manuscript until the
work on the MCM compatibility is finished. I'll change the MCM milestone
to 4.0.0.

**Comment by jimmielin** [comment]
Hi @yantosca, I guess it could be a 3.1.0 release feature since there aren't immediate requests for it, but I'll take a look at the MATLAB code this week to see if there's anything we need to change last minute. I'm actually not sure of the behavior of the MATLAB code in terms of `Update_RCONST()` - MATLAB uses the language's built-in integrators so there isn't a `ICNTRL` like the rest. It would be trivial to patch out these calls:

```matlab
function P = m_small_strato_Fun_Chem(T, Y) 
     
  global TIME FIX RCONST  
 
  Told = TIME;
  TIME = T;
  m_small_strato_Update_SUN;
  m_small_strato_Update_RCONST;
  
...
```

But I'm not sure how helpful that would be since it's in a tight loop and adding another argument and a `IF` check could slow things down. I suppose we could just add a note in the documentation about how to replicate the similar behavior, just move the two lines to another location where the constants would need to be updated.

**Comment by yantosca** [comment]
@jimmielin: Yeah, maybe we can make it a 3.1.0 update.  No need to rush, as long as the Fortran is doing what is advertised, I think.

**Comment by RolfSander** [comment]
I think for KPP-3.0.0 it would be sufficient to mention in the
documentation that the treatment of ICNTRL is somewhat different for the
generated C and Matlab codes:

https://kpp.readthedocs.io/en/latest/using_kpp/05_output_from_kpp.html#the-c-code

https://kpp.readthedocs.io/en/latest/using_kpp/05_output_from_kpp.html#the-matlab-code

**Comment by yantosca** [comment]
Doc now updated.

**Comment by jimmielin** [comment]
Thanks @yantosca! I think everything looks good now for the `3.0.0.rc.1` release, everything we've advertised in the manuscript is fully implemented. @RolfSander do we have any last minute changes we would like to make before Thursday? 

If we find any issues, etc. after release we can push out revised release candidate(s) as well while the manuscript is in review.

**Comment by RolfSander** [comment]
I also think that we're ready for 3.0.0-rc.1 now.

A few things we need to adjust before the release:

- Update KPP revision history in 00_revision_history.rst

- Update KPP_VERSION in gdata.h

- Update CHANGELOG.md

**Comment by yantosca** [comment]
Note: I pushed KPP 3.0.0-rc.0 but it didn't automatically generate a DOI.  I got this error:

```console
{
    "errors": "The license ID you have selected is not present in our system. For the available licenses please check in the following URL https://developers.zenodo.org/#licenses"
}
```
I will try to fix this in KPP 3.0.0-rc.1,

**Comment by RolfSander** [comment]
Thanks for creating the new version, @yantosca!

Regarding the version number, I think it should be 3.0.0-rc.0, not 3.0.0.rc-0 as in gdata.h and CHANGELOG.md.

**Comment by yantosca** [comment]
Just pushed KPP 3.0.0-rc.1 but it failed with not valid JSON error.  Ugh, I should have linted it!!!  Will fix.  Also will fix version number.

**Comment by yantosca** [comment]
Success! DOI now auto-generated  https://zenodo.org/record/6828026

**Comment by RolfSander** [comment]
Thanks for solving the problem with the auto-generated DOI, @yantosca!

**Comment by msl3v** [comment]
Removed GenerateComputeFamilies() and added a simple description of 
FEULER to branch dev. Pushed.
Let me know if anything else is needed.

**Comment by yantosca** [comment]
Hi @msl3v, I don't see your update in the dev branch.

**Comment by yantosca** [comment]
ah OK then.  Maybe it's an old commit.  Not sure why Github resurrected it.

Also FYI, we were waiting on Haipeng's paper to be reviewed.  It took the journal a while to assign reviewers so this is still ongoing.  Daniel wants 3.0.0 in GEOS-Chem 14.1.0.

**Comment by jimmielin** [comment]
Yes, hopefully JAMES will get back to us soon. It took about two months to find reviewers and it's been about 30 days in review so hopefully very soon :-)

**Comment by yantosca** [comment]
@jimmielin @RolfSander: Was speaking to Daniel, he was encouraging us to officially release KPP 3.0.0 soon so that we can start pulling in items for GEOS-Chem 14.1.0.  Would you mind if we start doing this?  (If we decide it's OK, I'll work on it the week of Oct 17th.  Rule 1 of Software Engineering: Never release new code on a Friday.)

**Comment by RolfSander** [comment]
Hmm, what if the reviewers want us to make some changes to the code? Should we then use version number 3.0.1 for the paper?

**Comment by yantosca** [comment]
@RolfSander: Daniel says that they probably won't request changes but you never know

**Comment by RolfSander** [comment]
Yes, I agree with Daniel that code change requests from the reviewers are quite unlikely. Still, I think it's good to be prepared. Besides, there may be more minor bugs like https://github.com/KineticPreProcessor/KPP/issues/60 in our current code. I think I prefer to continue with the "rc" version numbers for now. Would it be a limitation for GEOS-Chem 14.1.0 if the KPP version is called 3.0.0-rc.4?

**Comment by yantosca** [comment]
Thanks @RolfSander.  I am willing to wait until we get the reviews.  I think it's OK to have it be a release candidate version. GEOS-Chem 14.1.0 won't be released for a while so hopefully there's still time for a KPP 3.0.0 release before then.

**Comment by jimmielin** [comment]
Thanks @RolfSander for the reminder, I have updated the new DOI in the revised manuscript.

**Comment by yantosca** [comment]
I'll close out this issue as the 3.0.0 code is now released and the new DOI will be used going forward.  Thanks all!
